mrjob/dataproc.py

```diff
@@ -0,0 +1,873 @@
+        """
+        NOTE - mtai @ davidmarin - this function creates a "HadoopJob" to be passed to self._submit_hadoop_job
+
+        Reference...
+        https://cloud.google.com/dataproc/reference/rest/v1/projects.regions.jobs#HadoopJob
+
+        :param step_num:
+        :return:
+        """
+        args = list()
+        file_uris = list()
+        archive_uris = list()
+        properties = dict()
+
+        step = self._get_step(step_num)
+
+        assert step['type'] in ('streaming', 'jar'), 'Bad step type: %r' % (step['type'],)
+
+        # TODO - mtai @ davidmarin - Might be trivial to support jar running, see "mainJarFileUri" of variable "output_hadoop_job" in this function
+        #         https://cloud.google.com/dataproc/reference/rest/v1/projects.regions.jobs#HadoopJob
+        assert step['type'] == 'streaming', 'Jar not implemented'
+        main_jar_uri = _HADOOP_STREAMING_JAR_URI
+
+
+        # TODO - mtai @ davidmarin - Not clear if we should move _upload_args to file_uris, currently works fine as-is
+        args.extend(self._upload_args(self._upload_mgr))
+
+        # TODO - mtai @ davidmarin - Not clear if we should move some of these to properties dict to file_uris, currently works fine as-is
+        args.extend(self._hadoop_args_for_step(step_num))
+
+        mapper, combiner, reducer = (self._hadoop_streaming_commands(step_num))
+
+        if mapper:
+            args += ['-mapper', mapper]
+
+        if combiner:
+            args += ['-combiner', combiner]
+
+        if reducer:
+            args += ['-reducer', reducer]
+
+        for current_input_uri in self._step_input_uris(step_num):
+            args += ['-input', current_input_uri]
+
+        args += ['-output', self._step_output_uri(step_num)]
+
+        # TODO - mtai @ davidmarin - Add back support to specify a different jar URI
+        output_hadoop_job = {
+            "args": args,
+            "fileUris": file_uris,
+            "archiveUris": archive_uris,
+            "properties": properties,
+            "mainJarFileUri": main_jar_uri
+        }
+        return output_hadoop_job
+
+    def _launch_dataproc_cluster(self):
+        """Create an empty cluster on Dataproc, and set self._gcp_cluster to
+        its ID."""
+        self._wait_for_gcs_eventual_consistency()
+
+        self._gcp_cluster_start = time.time()
+
+        # make sure we can see the files we copied to GCS
+        log.info('Creating Dataproc Hadoop cluster')
+
+        # NOTE - mtai @ davidmarin - We create the cluster if its missing, otherwise we join an existing one
+        # NOTE - mtai @ davidmarin - The cluster MUST be created before submitting jobs, otherwise job submission fails
+
+        # TODO - mtai @ davidmarin - change the way we auto-create cluster names
+        # "clusterName must be a match of regex '(?:[a-z](?:[-a-z0-9]{0,53}[a-z0-9])?).'"
+        self._gcp_cluster = self._gcp_cluster or "default-%s" % self._opts['owner']
```

Comment:
Okay, having a single cluster named after `owner` is a terrible hack for a number of reasons. I was expecting the initial implementation would create a cluster named `mrjob-<random hex>`, and shut it down when the job is done (and then we could implemented something like pooling).

The fact the clusters can't shut themselves down (it's all up to `mrjob`) is a bit scary. I think it might be useful to add something like `mrjob/bootstrap/terminate_idle_cluster.sh` to the dataproc runner from the get-go. This script would almost work as is, but we could simplify it a bit: no `sudo`, assume `yarn` command exists, don't worry about how close we are to the end of a billing hour (since Google bills by the minute).

@evilsoapbox -- if you're listening, having Dataproc automatically terminate idle clusters would save your customers a lot of pain and would be a competitive advantage over Elastic MapReduce.